packages/view/lib/view.js

```diff
@@ -0,0 +1,383 @@
+        } else {
+          values = condition;
+        }
+
+        const ctx = on === 'post' || on === 'put' ? req.body : req.query;
+
+        if (
+          !every(values || {}, function(value, path) {
+            return value === true && path in ctx ? true : ctx[path] === value;
+          })
+        ) {
+          return this;
+        }
+      }
+
+      this.actionQueue.push(callback);
+    } else if (on === 'init') {
+      /* Init events are always fired in series, before any other actions
+       *
+       * Example:
+       *     view.on('init', function (next) {
+       *         // do something before any actions or queries have run
+       *     });
+       */
+
+      this.initQueue.push(callback);
+    } else if (on === 'render') {
+      /* Render events are always fired last in parallel, after any other actions
+       *
+       * Example:
+       *     view.on('render', function (next) {
+       *         // do something after init, action and query middleware has run
+       *     });
+       */
+
+      this.renderQueue.push(callback);
+    }
+
+    // TODO: Should throw if we didn't recognise the first argument!
+
+    return this;
+  }
+
+  /**
+   * Queues a mongoose query for execution before the view is rendered.
+   * The results of the query are set in `locals[key]`.
+   *
+   * Keys can be nested paths, containing objects will be created as required.
+   *
+   * The third argument `then` can be a method to call after the query is completed
+   * like function(err, results, callback), or a `populatedRelated` definition
+   * (string or array).
+   *
+   * Examples:
+   *
+   * view.query('books', keystone.list('Book').model.find());
+   *
+   *     an array of books from the database will be added to locals.books. You can
+   *     also nest properties on the locals variable.
+   *
+   * view.query(
+   *     'admin.books',
+   *      keystone.list('Book').model.find().where('user', 'Admin')
+   * );
+   *
+   *     locals.admin.books will be the result of the query
+   *     views.query().then is always called if it is available
+   *
+   * view.query('books', keystone.list('Book').model.find())
+   *     .then(function (err, results, next) {
+   *         if (err) return next(err);
+   *         console.log(results);
+   *         next();
+   *     });
+   *
+   * @api public
+   */
+
+  query(key, listOrKey, query, options) {
+    let list = listOrKey;
+    if (typeof listOrKey === 'string') {
+      list = this.keystone.getListByKey(listOrKey);
+    }
+    let locals = this.res.locals;
+    const parts = key.split('.');
+    const chain = new QueryCallbacks(options);
+
+    key = parts.pop();
+
+    for (let i = 0; i < parts.length; i++) {
+      if (!locals[parts[i]]) {
+        locals[parts[i]] = {};
+      }
+      locals = locals[parts[i]];
+    }
+
+    this.queryQueue.push(async () => {
+      const callbacks = chain.callbacks;
+      const context = this.keystone.getAccessContext('admin', this.req);
```

Comment:
is `admin` always going to be schemaName?

Reply:
`admin` will be the default schema name used when creating the apollo server used by the Admin UI. At some point this will probably become configurable. Also, we will soon be able to spin up multiple apollo servers with different schemas, each of which will be accessible via a different schema name. As such, it probably makes sense to make the `schemaName` a config option for the view, with a default of `"admin"`.

Reply:
adding this as part of View constructor now.